pkg/proxy/middleware.go

```diff
@@ -0,0 +1,77 @@
+package proxy
+
+import (
+	"bytes"
+	"log"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+var _ http.Handler = &RetryMiddleware{}
+
+type RetryMiddleware struct {
+	nextHandler http.Handler
+	MaxRetries  int
+	rSource     *rand.Rand
+}
+
+func NewRetryMiddleware(maxRetries int, other http.Handler) *RetryMiddleware {
+	if maxRetries < 1 {
+		panic("invalid retries")
+	}
+	return &RetryMiddleware{
+		nextHandler: other,
+		MaxRetries:  maxRetries,
+		rSource:     rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	var capturedResp *responseBuffer
+	for i := 0; ; i++ {
+		capturedResp = &responseBuffer{
+			header: make(http.Header),
+			body:   bytes.NewBuffer([]byte{}),
+		}
+		// call next handler in chain
+		r.nextHandler.ServeHTTP(capturedResp, request.Clone(request.Context()))
+
+		if i == r.MaxRetries || // max retries reached
+			request.Context().Err() != nil || // abort early on timeout, context cancel
+			capturedResp.status != http.StatusBadGateway &&
+				capturedResp.status != http.StatusServiceUnavailable {
```

Comment:
Should we include `StatusGatewayTimeout`?

Reply:
Yes, I think `408`, `502`, `503`, `504` make sense. Perhaps a helper func `isRetryableStatusCode(int) bool`?